attribute_based.py

```diff
@@ -0,0 +1,133 @@
+    img_center = np.array([.5, .5])
+
+
+    info = pickle.load(open('util_files/places_scene_info.pkl', 'rb'))
+    idx_to_scene = info['idx_to_scene']
+    idx_to_scenegroup = info['idx_to_scenegroup']
+    sceneidx_to_scenegroupidx = info['sceneidx_to_scenegroupidx']
+
+    if FACE_DETECT == 0:
+        client = boto3.client('rekognition')
+        if os.path.exists('{}_rekognitioninfo.pkl'.format(args.folder)):
+            detect_info = pickle.load(open('{}_rekognitioninfo.pkl'.format(args.folder), 'rb'))
+        else:
+            detect_info = {}
+    elif FACE_DETECT == 1:
+        cascPath = "haarcascade_frontalface_default.xml"
+        #Had to add in haarcascades location
+        faceCascade = cv2.CascadeClassifier(cv2.data.haarcascades + cascPath)
+
+    for i, (data, target) in enumerate(tqdm(dataloader)):
+
+        #We assume attribute information is in 1st index
+        attribute = target[1]
+
+        # Only look at image if there is an attribute to analyze (Note attribute require a bbox around the person or thing to analyze)
+        if len(attribute)> 1:
+            shape = list(data.size())[1:]
+            bbox = attribute[1]
+            bbox_adjust = np.array([bbox[0]*shape[1], bbox[1]*shape[1], bbox[2]*shape[0], bbox[3]*shape[0]])
+            pixel_size = (bbox_adjust[1]-bbox_adjust[0])*(bbox_adjust[3]-bbox_adjust[2])
+            size = (bbox[1]-bbox[0])*(bbox[3]-bbox[2])
+            person_center = np.array([bbox[0] + (bbox[1]/2.), bbox[2] + (bbox[3]/2.)])
+
+            distance = np.linalg.norm(person_center - img_center)
+            pic = data.data.cpu().numpy()
+
+            if FACE_DETECT == 0:
+                detect_info = detect_face(target[3], detect_info, client)
+                faceDetails = detect_info[target[3]]['FaceDetails']
+                if len(detect_info) % 20 == 0:
+                    pickle.dump(detect_info, open('{}_rekognitioninfo.pkl'.format(args.folder), 'wb'))
+
+                yes_face = False
+                for face in faceDetails:
+                    if face['Confidence'] > .9:
+                        yes_face = True
+            elif FACE_DETECT == 1:
+                
+                image = cv2.imread(target[3])
+                gray = cv2.cvtColor(image, cv2.COLOR_RGB2GRAY)
+                faces = faceCascade.detectMultiScale(
+                   gray,
+                   scaleFactor=1.1,
+                   minNeighbors=5,
+                   #minSize=(30, 30),
+                   flags = cv2.CASCADE_SCALE_IMAGE
+                )
+                yes_face = False
+                if len(faces) > 0:
+                    yes_face = True
+
+            # If there's no face detected and the person is too small, attribute cannot be distinguished
+            if not yes_face or pixel_size < 1000.:
+                scene_group = target[4]
+
+                if not yes_face:
+                    no_faces[attribute[0]-1].append((size, pixel_size, scene_group))
```

Comment:
it was only gender[0]-1 before, because the labels for datasets.py were reversed before (it asked for [male, female], but then the notebook treats it as [female, male], or something like that). it should not be attribute[0]-1 here, it should just be attribute[0] (also the lines below).